File: backend/app/models/finance.py
"""财务域：订单、收支、课消、账户充值。"""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.core.db import Base
from app.core.timeutil import now as _utcnow

class FinanceOrder(Base):
    """财务订单（报名/续费/转课/退课/充值等）。"""

    __tablename__ = "finance_orders"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    order_no: Mapped[str] = mapped_column(String(32), unique=True, index=True)
    student_id: Mapped[int] = mapped_column(ForeignKey("students.id"), index=True)
    # enroll | renew | transfer | drop | recharge | refund | other
    order_type: Mapped[str] = mapped_column(String(32), default="enroll", index=True)
    item_summary: Mapped[str] = mapped_column(Text, default="")
    # JSON courses snapshot
    courses: Mapped[str] = mapped_column(Text, default="[]")
    receivable: Mapped[float] = mapped_column(Float, default=0.0)
    received: Mapped[float] = mapped_column(Float, default=0.0)
    arrears: Mapped[float] = mapped_column(Float, default=0.0)
    # paid | partial | unpaid | void
    status: Mapped[str] = mapped_column(String(32), default="paid", index=True)
    source: Mapped[str] = mapped_column(String(64), default="机构创建")
    performance_owner_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)
    handler_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)
    enrollment_id: Mapped[int | None] = mapped_column(
        ForeignKey("enrollment_records.id"), nullable=True, index=True
    )
    pay_method: Mapped[str] = mapped_column(String(64), default="")
    handled_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True, index=True)
    paid_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    created_by: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=_utcnow)

class FinanceTransaction(Base):
    """收支明细。"""

    __tablename__ = "finance_transactions"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    handled_at: Mapped[datetime] = mapped_column(DateTime, default=_utcnow, index=True)
    # 报名/续费 | 账户充值 | 退费 | …
    item: Mapped[str] = mapped_column(String(64), default="报名/续费")
    # income | expense
    tx_type: Mapped[str] = mapped_column(String(16), default="income", index=True)
    # pending | confirmed | void
    status: Mapped[str] = mapped_column(String(32), default="pending", index=True)
    amount: Mapped[float] = mapped_column(Float, default=0.0)
    pay_method: Mapped[str] = mapped_column(String(64), default="")
    account: Mapped[str] = mapped_column(String(64), default="")
    handler_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)
    order_id: Mapped[int | None] = mapped_column(ForeignKey("finance_orders.id"), nullable=True, index=True)
    student_id: Mapped[int | None] = mapped_column(ForeignKey("students.id"), nullable=True, index=True)
    payer_name: Mapped[str] = mapped_column(String(128), default="")
    voucher: Mapped[str] = mapped_column(String(128), default="")
    flow_no: Mapped[str] = mapped_column(String(128), default="")
    remark: Mapped[str] = mapped_column(Text, default="")
    created_by: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=_utcnow)
    confirmed_at: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    confirmed_by: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)

class CourseConsumption(Base):
    """课消记录（点名产生）。"""

    __tablename__ = "course_consumptions"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    student_id: Mapped[int] = mapped_column(ForeignKey("students.id"), index=True)
    class_id: Mapped[int | None] = mapped_column(ForeignKey("class_rooms.id"), nullable=True, index=True)
    course_id: Mapped[int | None] = mapped_column(ForeignKey("courses.id"), nullable=True, index=True)
    record_id: Mapped[int | None] = mapped_column(ForeignKey("class_records.id"), nullable=True, index=True)
    teacher_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)
    # hour_consume | …
    consume_type: Mapped[str] = mapped_column(String(32), default="课时课消")
    # roll_call | manual
    source: Mapped[str] = mapped_column(String(32), default="点名")
    hours: Mapped[float] = mapped_column(Float, default=0.0)
    amount: Mapped[float] = mapped_column(Float, default=0.0)
    # JSON: [{"package_id": 1, "hours": 1.0, "amount": 200.0}]
    package_allocations: Mapped[str] = mapped_column(Text, default="[]")
    # 课包余额不足时仍保留的欠课时，供续费与课时处理追踪。
    uncovered_hours: Mapped[float] = mapped_column(Float, default=0.0)
    consumed_at: Mapped[datetime] = mapped_column(DateTime, default=_utcnow, index=True)
    # normal | void
    status: Mapped[str] = mapped_column(String(32), default="normal", index=True)
    created_by: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=_utcnow)

class StudentAccount(Base):
    """学员账户余额。"""

    __tablename__ = "student_accounts"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    student_id: Mapped[int] = mapped_column(ForeignKey("students.id"), unique=True, index=True)
    balance: Mapped[float] = mapped_column(Float, default=0.0)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=_utcnow, onupdate=_utcnow)

class RechargeRecord(Base):
    """账户充值记录。"""

    __tablename__ = "recharge_records"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    student_id: Mapped[int] = mapped_column(ForeignKey("students.id"), index=True)
    amount: Mapped[float] = mapped_column(Float, default=0.0)
    balance_after: Mapped[float] = mapped_column(Float, default=0.0)
    pay_method: Mapped[str] = mapped_column(String(64), default="微信")
    handler_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)
    order_id: Mapped[int | None] = mapped_column(ForeignKey("finance_orders.id"), nullable=True)
    # success | void
    status: Mapped[str] = mapped_column(String(32), default="success")
    remark: Mapped[str] = mapped_column(Text, default="")
    created_at: Mapped[datetime] = mapped_column(DateTime, default=_utcnow, index=True)

class OrderOperationLog(Base):
    """订单操作日志。"""

    __tablename__ = "order_operation_logs"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    order_id: Mapped[int] = mapped_column(ForeignKey("finance_orders.id"), index=True)
    # create | view | print_receipt | void | update | other
    action: Mapped[str] = mapped_column(String(32), default="other", index=True)
    action_label: Mapped[str] = mapped_column(String(64), default="")
    detail: Mapped[str] = mapped_column(Text, default="")
    operator_id: Mapped[int | None] = mapped_column(ForeignKey("users.id"), nullable=True)
    operator_name: Mapped[str] = mapped_column(String(128), default="")
    created_at: Mapped[datetime] = mapped_column(DateTime, default=_utcnow, index=True)

File: frontend/src/api/finance.ts
import client from './client'
import { asPageResult, type PageResult } from './paging'

export interface OrderLineItem {
  course_id?: number | null
  name: string
  tag: string
  type?: string
  price_label: string
  quantity_label: string
  hours?: number
  unit_price?: number
  total: number
  gift_qty: string
  leave_free: string
  discount: number
  subtotal: number
  coupon: string
  receivable: number
  class_name: string
  valid_until: string
}

export interface FinanceTransaction {
  id: number
  handled_at?: string
  item: string
  type: string
  tx_type: string
  status: string
  status_code: string
  amount: number
  pay_method: string
  account: string
  handler: string
  order_id?: number | null
  order_no: string
  student_id?: number | null
  payer: string
  voucher: string
  flow_no: string
  remark: string
  created_at?: string
}

export interface FinanceOrder {
  id: number
  order_no: string
  student_id: number
  student: string
  phone: string
  order_type: string
  order_type_label: string
  item: string
  item_summary: string
  courses: { id?: number; name?: string; type?: string; price_label?: string; hours?: number; unit_price?: number }[]
  receivable: number
  received: number
  arrears: number
  status: string
  status_label: string
  source: string
  performance_owner: string
  handler: string
  pay_method: string
  handled_at?: string | null
  paid_at?: string | null
  creator: string
  created_at?: string
}

export interface FinanceOrderDetail extends FinanceOrder {
  student_grade?: string
  student_school?: string
  parent_name?: string
  gender?: string
  internal_notes?: string
  external_notes?: string
  line_items: OrderLineItem[]
  line_receivable_sum: number
  payments: FinanceTransaction[]
  performance_owner_id?: number | null
  handler_id?: number | null
  enrollment_id?: number | null
}

export interface TransactionListResult extends PageResult<FinanceTransaction> {
  summary?: {
    income: number
    pending_income: number
    expense: number
    pending_expense: number
  }
}

export interface CourseConsumption {
  id: number
  consumed_at?: string
  student_id: number
  student: string
  student_phone?: string
  student_grade?: string
  class_id?: number | null
  class_name: string
  course_id?: number | null
  course_name: string
  course_type?: string
  course_type_label?: string
  course_grade?: string
  subject?: string
  term?: string
  teacher: string
  teacher_id?: number | null
  /** 该课次全部上课老师 id（与点名/排课一致） */
  teacher_ids?: number[]
  record_id?: number | null
  consume_type: string
  source: string
  hours: number
  hours_label: string
  uncovered_hours: number
  amount: number
  status: string
  status_label?: string
  created_at?: string
}

export interface CourseConsumptionOrder {
  package_id?: number | null
  order_id?: number | null
  order_no: string
  course_name: string
  unit_price: number
  hours: number
  gift_hours: number
  amount: number
}

export interface CourseConsumptionDetail extends CourseConsumption {
  class_start?: string | null
  class_end?: string | null
  roll_at?: string | null
  attendance_status?: string
  attendance_status_label?: string
  operator?: string
  operation_time?: string | null
  orders: CourseConsumptionOrder[]
}

export interface ConsumptionListResult extends PageResult<CourseConsumption> {
  summary?: { amount: number }
}

export interface RechargeRecord {
  id: number
  student_id: number
  student: string
  phone: string
  amount: number
  balance: number
  pay_method: string
  handler: string
  status: string
  remark: string
  created_at?: string
}

export interface IncomeReport {
  pending_income: number
  confirmed_income: number
  total_income: number
  by_pay_method: { method: string; count: number; amount: number }[]
  income_chart?: { date: string; pending: number; confirmed: number; total: number }[]
  course_consumption?: {
    total_hours: number
    total_amount: number
    total_count: number
    chart?: { date: string; hours: number; amount: number; count: number }[]
    by_course: {
      course_name: string
      course_type?: string
      course_type_label?: string
      count: number
      hours: number
      amount: number
    }[]
  }
}

export interface PendingHoursMetrics {
  package_count: number
  purchased_hours: number
  gift_hours: number
  total_hours: number
  consumed_hours: number
  pending_hours: number
  pending_value: number
  expired_hours: number
  expiring_soon_hours: number
  consumption_rate: number
  risk_status: 'expired' | 'expiring' | 'normal' | 'consumed'
}

export interface PendingHoursSummary extends PendingHoursMetrics {
  student_count: number
  pending_student_count: number
}

export interface PendingHoursCourse extends PendingHoursMetrics {
  course_id: number
  course_name: string
  course_type: string
  course_type_label: string
  student_count: number
}

export interface PendingHoursItem extends PendingHoursMetrics {
  student_id: number
  student_name: string
  student_phone: string
  student_grade: string
  student_status: string
  course_id: number
  course_name: string
  course_type: string
  course_type_label: string
  valid_until?: string | null
}

export interface PendingHoursReport {
  as_of: string
  summary: PendingHoursSummary
  by_course: PendingHoursCourse[]
  items: PendingHoursItem[]
}

export async function listOrdersApi(params: {
  order_no?: string
  student_q?: string
  order_type?: string
  page?: number
  page_size?: number
} = {}): Promise<PageResult<FinanceOrder>> {
  const res = await client.get('/finance/orders', { params })
  return asPageResult<FinanceOrder>(res.data.data, params.page ?? 1, params.page_size ?? 20)
}

export async function createOrderApi(payload: {
  student_id: number
  order_type?: string
  item_summary?: string
  receivable?: number
  received?: number
  pay_method?: string
  handled_at?: string | null
  performance_owner_id?: number | null
}): Promise<FinanceOrder> {
  const res = await client.post('/finance/orders', payload)
  return res.data.data
}

export async function getOrderDetailApi(
  orderId: number,
  opts?: { logView?: boolean },
): Promise<FinanceOrderDetail> {
  const res = await client.get(`/finance/orders/${orderId}`, {
    params: opts?.logView ? { log_view: true } : undefined,
  })
  return res.data.data
}

export async function voidOrderApi(orderId: number): Promise<FinanceOrderDetail> {
  const res = await client.post(`/finance/orders/${orderId}/void`)
  return res.data.data
}

export interface OrderOperationLog {
  id: number
  order_id: number
  action: string
  action_label: string
  detail: string
  operator_id?: number | null
  operator_name: string
  created_at?: string
}

export async function listOrderLogsApi(
  orderId: number,
  params: { page?: number; page_size?: number } = {},
): Promise<{ items: OrderOperationLog[]; total: number; page?: number; page_size?: number }> {
  const res = await client.get(`/finance/orders/${orderId}/logs`, {
    params: {
      page: params.page ?? 1,
      page_size: params.page_size ?? 50,
    },
  })
  return res.data.data
}

/** 下载订单收据 PDF */
export async function downloadOrderReceiptApi(
  orderId: number,
  meta?: { orderNo?: string; studentName?: string },
): Promise<void> {
  const res = await client.get(`/finance/orders/${orderId}/receipt`, {
    responseType: 'blob',
    timeout: 60_000,
  })
  const blob = res.data as Blob
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  const student = (meta?.studentName || '学员').replace(/[\\/:*?"<>|]/g, '-').trim() || '学员'
  const no = (meta?.orderNo || String(orderId)).replace(/[\\/:*?"<>|]/g, '-').trim()
  a.download = `收据_${student}_${no}.pdf`
  document.body.appendChild(a)
  a.click()
  a.remove()
  URL.revokeObjectURL(url)
}

export async function listTransactionsApi(params: {
  item?: string
  tx_type?: string
  status?: string
  include_void?: boolean
  page?: number
  page_size?: number
} = {}): Promise<TransactionListResult> {
  const res = await client.get('/finance/transactions', { params })
  const data = res.data.data || {}
  const page = asPageResult<FinanceTransaction>(data, params.page ?? 1, params.page_size ?? 20)
  return { ...page, summary: data.summary }
}

export async function confirmTransactionsApi(ids: number[]): Promise<{ confirmed: number }> {
  const res = await client.post('/finance/transactions/confirm', { ids })
  return res.data.data
}

export async function voidTransactionApi(id: number): Promise<FinanceTransaction> {
  const res = await client.post(`/finance/transactions/${id}/void`)
  return res.data.data
}

export async function listConsumptionsApi(params: {
  student_q?: string
  class_id?: number
  course_id?: number
  course_type?: string
  teacher_id?: number
  consume_type?: string
  source?: string
  status?: string
  hide_void?: boolean
  start_date?: string
  end_date?: string
  grade?: string
  subject?: string
  term?: string
  page?: number
  page_size?: number
} = {}): Promise<ConsumptionListResult> {
  const res = await client.get('/finance/consumptions', { params })
  const data = res.data.data || {}
  const page = asPageResult<CourseConsumption>(data, params.page ?? 1, params.page_size ?? 20)
  return { ...page, summary: data.summary }
}

export async function getConsumptionDetailApi(id: number): Promise<CourseConsumptionDetail> {
  const res = await client.get(`/finance/consumptions/${id}`)
  return res.data.data
}

export async function listRechargesApi(params: {
  student_q?: string
  page?: number
  page_size?: number
} = {}): Promise<PageResult<RechargeRecord>> {
  const res = await client.get('/finance/recharges', { params })
  return asPageResult<RechargeRecord>(res.data.data, params.page ?? 1, params.page_size ?? 20)
}

export async function createRechargeApi(payload: {
  student_id: number
  amount: number
  pay_method?: string
  remark?: string
}): Promise<RechargeRecord> {
  const res = await client.post('/finance/recharges', payload)
  return res.data.data
}

export async function getIncomeReportApi(params: {
  start_date?: string
  end_date?: string
} = {}): Promise<IncomeReport> {
  const res = await client.get('/finance/income-report', { params })
  return res.data.data
}

export async function getPendingHoursReportApi(): Promise<PendingHoursReport> {
  const res = await client.get('/finance/pending-hours-report')
  return res.data.data
}
